app/services/course/assessment/submission/auto_grading_service.rb

```diff
@@ -25,10 +29,22 @@ def grade(submission)
   private
 
   # Grades the answers in the provided submission.
+  #
+  # Retries are implemented in the case where a race condition occurs, ie. when a new
+  # attempting answer is created after the submission is finalised, but before the
+  # autograding job is run for the submission.
   def grade_answers(submission)
-    jobs = submission.latest_answers.map { |answer| grade_answer(answer) }
-    wait_for_jobs(jobs)
-    aggregate_failures(jobs.map { |job| job.job.reload })
+    jobs_by_qn = submission.latest_answers.map { |a| [a.question_id, grade_answer(a)] }.to_h
+    wait_for_jobs(jobs_by_qn.values)
+
+    tries = 0
+    while submission.reload.latest_answers.reject(&:graded?).any? && tries <= MAX_RETRIES
```

Comment:
Just realized there are evaluated answers, should reject them as well. Or filter answers by attempting and submitted state.